### PR TITLE
Fixes react state race condition for field array methods

### PR DESCRIPTION
**Pull request by revich2**
This PR solves issues from this topics - #6037, #6039

After deep research i found problem, why this strange behaviour happens. It is race condition of react state updates.
You can see issue here - https://codesandbox.io/s/delicate-flower-rgzq0?file=/src/App.js

1) If you will use external RHForm for adding new items to field array, then you will see collision - items will be added incorrect.
2) If you will use native html form for adding new items to field array, then all will be OK.

It really strange because i understood that it is race condition of react state updates, but i don't know that causes that.

All integration tests have done.

I have problem with unit tests of this PR, because some tests, which contains `useFieldArray`, fails due to `setTimeout(..., 0)`. which i used in `createFormControl` to schedule `_subjects.state.next` as early as it possible.
And i don't know how i can fix failed unit test, i tried to use `jest.useFakeTimers`, but it didn't help.

If who knows how to solve problem with `setTimeout(..., 0)` in unit tests, then i will be able to fix them.

@bluebill1049 check it, please

### CODE REVIEW

**Comment by bluebill1049** [comment]
thanks for the PR and investigation. the codesandbox is a lot easier for me to disgest as well. please refer to the doc on the `reset`.

https://react-hook-form.com/api/useform/reset/

Screen Shot 2021-08-06 at 12 25 21 pm

**Comment by bluebill1049** [comment]
also worth mentioning, handleSubmit onSubmit is async not sync.

**Comment by revich2** [comment]
@bluebill1049 oh, it is unobviously thing, really.

what do you think about this PR? all integration tests have done and this logic works obviously

will it be merged? after fixes unit tests, of course

**Comment by bluebill1049** [comment]
> @bluebill1049 oh, it is unobviously thing, really.

I guess it's a question of whether we support async onSubmit for users, I guess the answer is pretty simple: 'yes' 

-> onSubmit -> async -> setState

vs

-> onSubmit -> setState

**Comment by revich2** [comment]
@bluebill1049 but what the problem with async onSubmit and changes from this PR? sorry, i'm not understand you will approve this changes or not

**Comment by bluebill1049** [comment]
I will not have `setTimeout` in this lib, if you have a strong opinion on the `reset` API, please fork the repo and publish a version of your own.

**Comment by revich2** [comment]
@bluebill1049 this changes not about `reset` API, it about field array methods. `reset` isn't influence on this case
while i was investigating this case i found that `control._bathFieldArrayUpdate` has to be before `setFields` in field array methods functions, because it's also influence on race condition and it without any `setStates`.
Screenshot 2021-08-06 at 10 25 28 AM

with `reset` i agree with you, but array methods can used only in external RHForm, not in onSubmit their form

File: src/logic/createFormControl.ts [hunk 1]
```diff
@@ -706,12 +706,16 @@ export function createFormControl<
       unsetEmptyArray(_formState.dirtyFields, name);
     }
 
-    _subjects.state.next({
-      isDirty: _getIsDirty(name, omitKey(updatedFieldArrayValues, keyName)),
-      dirtyFields: _formState.dirtyFields,
-      errors: _formState.errors,
-      isValid: _formState.isValid,
-    });
+    setTimeout(
```
Comment (bluebill1049) [inline src/logic/createFormControl.ts:709]: what's this?

File: src/useFieldArray.ts [hunk 2]
```diff
@@ -118,6 +113,15 @@ export const useFieldArray = <
...
 
+    setFields(
+      mapIds(
+        updatedFieldArrayValues as Partial<
+          FieldArray<TFieldValues, TFieldArrayName>
+        >[],
+        keyName,
+      ),
+    );
+
```
Comment (bluebill1049) [inline src/useFieldArray.ts:124]: I don't mind those changes, as long as there is a test case to prove this does fix an issue.

**Comment by bluebill1049** [comment]
i will take over PR and remove the `setTimeout`, your effort is appreciated here.

File: src/logic/createFormControl.ts (hunk 1 above)
Comment (bluebill1049) [reply src/logic/createFormControl.ts]: This is not accepted.

**Pull request merged**